[Style] Convert grid-auto-columns/grid-auto-rows to use strong style types

weinig opened the pull request:
#### 3bc1b8c9336ae14879f83e0c9e2a8da048fe8621
<pre>
[Style] Convert grid-auto-columns/grid-auto-rows to use strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=295866">https://bugs.webkit.org/show_bug.cgi?id=295866</a>

Reviewed by Darin Adler.

Adds a new type, `Style::GridTrackSizes` to represent the `grid-auto-columns/grid-auto-rows`
properties. To make it work, `Style::GridTrackSize` was updated to work with
the style type system and given explicit types to represent the `fit-content`
and `min-max` states.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleGridData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/grid/StyleGridTemplateList.cpp:
* Source/WebCore/style/values/grid/StyleGridTemplateList.h:
* Source/WebCore/style/values/grid/StyleGridTrackSize.cpp:
* Source/WebCore/style/values/grid/StyleGridTrackSize.h:
* Source/WebCore/style/values/grid/StyleGridTrackSizes.cpp: Added.
* Source/WebCore/style/values/grid/StyleGridTrackSizes.h: Added.

Canonical link: <a href="https://commits.webkit.org/297352@main">https://commits.webkit.org/297352@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e028ac40207b761a280033936e5011540d963371

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111472 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31138 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21603 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117504 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113434 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39720 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/84717 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114419 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25411 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100346 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65165 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24759 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18481 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61326 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94797 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18547 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120727 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38521 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28628 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/93643 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38897 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96614 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93469 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23811 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38578 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16354 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34558 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38410 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43887 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38075 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41408 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39777 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/style/values/grid/StyleGridTrackSize.h:
```diff
@@ -142,10 +224,19 @@ struct GridTrackSize {
+    Type m_type;
+    GridTrackBreadth m_minTrackBreadth;
+    GridTrackBreadth m_maxTrackBreadth;
+    GridTrackFitContentLength m_fitContentTrackLength;
```
Seems good style to always initialize as long as it doesn’t have any cost. Can then remove constructor initializations that set members to those defaults.
```suggestion
    Type m_type { Type::Breadth };
    GridTrackBreadth m_minTrackBreadth { CSS::Keyword::Auto { } };
    GridTrackBreadth m_maxTrackBreadth { CSS::Keyword::Auto { } };
    GridTrackFitContentLength m_fitContentTrackLength { 0_css_px };
```

darinadler commented on Source/WebCore/style/values/grid/StyleGridTrackSizes.cpp:
```diff
@@ -0,0 +1,45 @@
+    std::call_once(onceFlag, [] {
+        staticValue.construct(GridTrackSize { CSS::Keyword::Auto { } });
+    });
+    return staticValue.get();
```
Is there really a need to call this from multiple threads? This idiom seems overkill if not.

weinig replied:
There isn't. I got confused about which NeverDestroyed to use because I couldn't remember what the rules for static initializers in a function were. Maybe `LazyNeverDestroyed` needs a more clear name to remind people (like me) about when to use it.

weinig committed:
[Style] Convert grid-auto-columns/grid-auto-rows to use strong style types
https://bugs.webkit.org/show_bug.cgi?id=295866

Reviewed by Darin Adler.

Adds a new type, `Style::GridTrackSizes` to represent the `grid-auto-columns/grid-auto-rows`
properties. To make it work, `Style::GridTrackSize` was updated to work with
the style type system and given explicit types to represent the `fit-content`
and `min-max` states.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleGridData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/grid/StyleGridTemplateList.cpp:
* Source/WebCore/style/values/grid/StyleGridTemplateList.h:
* Source/WebCore/style/values/grid/StyleGridTrackSize.cpp:
* Source/WebCore/style/values/grid/StyleGridTrackSize.h:
* Source/WebCore/style/values/grid/StyleGridTrackSizes.cpp: Added.
* Source/WebCore/style/values/grid/StyleGridTrackSizes.h: Added.

Canonical link: https://commits.webkit.org/297352@main

rkirsling commented on Source/WebCore/style/values/grid/StyleGridTrackSize.cpp:
```diff
@@ -128,42 +74,25 @@ auto Blending<GridTrackSize>::blend(const GridTrackSize& from, const GridTrackSi
 }
-// MARK: - Logging
-
-TextStream& operator<<(TextStream& ts, const GridTrackSize& value)
```
The PS Debug build was broken (at CSSValueAggregates.h) by this being removed:
https://build.webkit.org/#/builders/65/builds/54197/steps/9/logs/stdio

Is it meant to come from somewhere else instead?

rkirsling replied:
Restoring the function in https://github.com/WebKit/WebKit/pull/48036; feel free to replace with a different fix.
